### ISSUE TITLE
Repeated characters when --delay is large

### ISSUE THREAD

**Issue by jidanni** [jordansissel/xdotool#423]
On my system today the repeating starts at around 996 ms.
for i in 980 990 998; do xdotool type --delay $i $i.; done
980.990.999988..
Even scarier is when one tries e.g., --delay 4000...
Also key --delay has similar problems.

xdotool version 3.20160805.1

**Comment by jordansissel** [comment]
I’m on my phone so it’s hard to check the code, but my guess is that xdotool splits the delay between the key press and key release, and maybe it shouldn’t do that for the reasons you’ve found!

Assuming I’m remembering the code right, we should be able to make it so the delay between press and release is much shorter

**Comment by jidanni** [comment]
Well, apparently there is some number hardwired into the program that the user cannot access.

Therefore, whatever you change, be sure to expose it so the user can also adjust it, in case whatever is assumed causes problems on his system.

**Comment by limitedAtonement** [comment]
I just did some experimentation which seems to verify

> xdotool splits the delay between the key press and key release

(I think "apparently there is some number hardwired into the program that the user cannot access." is unrelated.) It appears that the key down event is issued immediately, and if `--delay` is specified, the key up event is done half way through the delay. So, a `xdotool type --delay 2000 ss;` will

* 0ms: s keydown
* 1000 ms: s key up
* 2000 ms: s key down
* 3000 ms: s key up
* 4000 ms: program quits with exit code 0

In other words, s is held for a second, released for a second, then held again for a second and released for a second. If your X server (window manager? I'm not sure what handles this behavior) repeats letters that are held for a long time like mine does, you'll get repeated keys.

**Comment by jidanni** [comment]
That's a great explanation to add to the man page. However I would use two different characters instead of the two of the same.

Also the man page should mention how to fine tune this behavior too avoid the repeats. For instance the length of the time the key stays down... how to adjust that etc.

**Comment by limitedAtonement** [comment]
> ...be sure to expose the raw controls too in case some users need finer grained control.

I think we have the finer-grained controls in the form of `keydown` and `keyup` (with sleeps).

**Comment by limitedAtonement** [comment]
> I agree this behavior is not desirable. I think we can change xdotool to keep it from causing key repeats by changing the code to be a bit wiser -- reducing the delay used in the "down" event and spending most of the time waiting after the "up"

It's nice to know what it's doing. Predictable is better than unknown. Better still is configurable. Changing to a fixed 50ms delay or whatever is actually probably not good. Right now, we can at least adjust the "key down" time, but with a fixed key down time, it will no longer be adjustable. While you're there, maybe a "**dwell time ratio**" setting would be good? So the current behavior could be preserved by specifying a "dwell time ratio" of `0.5` (which should probably be the default). A ratio of `0.1` would mean "spend 10% of the delay time with key down", etc. Of course, additionally an absolute "**dwell time**" setting would be nice in milliseconds, too. I'm not sure if this is over-engineered, but it seems good to me.

**Comment by jidanni** [comment]
Wow, with all controls exposed, any "are you a human" typing test can be beat!

**Comment by Dunkhan** [comment]
So just to be clear, as it currently stands there is no way for me to for example 
wait 5 seconds and then send key a down, key a up immediately
could we get another delay modifier like --sleep or --wait that simply passes some time before sending the keystrokes?

**Comment by limitedAtonement** [comment]
> So just to be clear, as it currently stands there is no way for me to for example
> wait 5 seconds and then send key a down, key a up

Maybe not in one `xdotool` command, but something like this should work:

```bash
sleep 5;
xdotool keydown a keyup a || exit;
```

**Comment by jordansissel** [comment]
> wait 5 seconds and then send key a down, key a up

@Dunkhan xdotool has a `sleep` command which may help if you need it.

I'll take a look at #423 which should help resolve this issue

**Comment by jordansissel** [comment]
#471 is merged which should help with this issue (character repetition when delay is large). It solves this by having xdotool hold the "key down" event for 50ms before releasing each key. The delay now affects only the time between each keystroke (down + up)

**Comment by jordansissel** [comment]
> The pattern is: CTRL+<client-number> CTRL+t (search box) type , enter, etc. There must be a delay between switching clients and navigating to the search box.

@limitedAtonement I wonder if #471 affects you. The change in #471 only changes how the `type` command uses `--delay` and does not change the `key` `--delay` flag behavior. Please let me know if something is broken by this change in a new issue where I'd also like to hear about your use case a bit more :)

**Comment by jordansissel** [comment]
> send alt key down, then press tab key, wait 3 seconds then send alt key up

@Dunkhan Try this:

```
xdotool keydown Alt keydown Tab sleep 3 keyup Alt
```

Keep in mind that the "Tab" key will still be pressed when xdotool ends. 

If you'd rather "Alt+Tab" be held together for 3 seconds and then released, you can do this:

```
xdotool keydown Alt_L+Tab sleep 3 keyup Alt+Tab
```

**Comment by limitedAtonement** [comment]
> I'd also like to hear about your use case a bit more

Thank you for taking note. I worked around my problem by incorporating calls to `sleep` (outside xdotool). I think having a fixed 50 millisecond dwell time is an excellent improvement. I have nothing to add about my usecase beyond what's written above.
